### PR TITLE
Agent runtime features

### PR DESCRIPTION
**Pull request by asherk7**
## Description

Implements all six runtime items from issue #88 section 7: retry/timeout around
model calls, context window management, streaming chat output, observability
hooks, workspace schema versioning, and SQLite connection lifecycle.

Part of #88 (does not close it — other sections are owned separately)

## What changed

**Retry/timeout around model calls**
- `Minion` gained `request_timeout` (default 60s per HTTP request), `max_retries`
  (default 2) and `retry_base_delay`. Transient provider errors (HTTP 429/5xx,
  connection failures, timeouts) retry with exponential backoff; auth and other
  4xx errors still fail immediately.

**Streaming output in chat**
- New `Minion.run_stream()` yields the reply as text deltas; `chat start` now
  prints them as they arrive instead of waiting for the full completion.
  If a stream fails mid-turn, the partial reply is persisted with an error
  marker so the transcript matches what was shown on screen.

**Context window management**
- Chat history sent to the model is capped at the newest 40 messages
  (`trim_message_history`), cutting only at user-turn boundaries so resumed
  sessions never start mid-tool-exchange. The on-disk JSONL transcript stays
  complete; long-term recall remains the distiller's job.

**Observability hooks + `--verbose`**
- Optional `on_tool_call(name, args)` and `on_turn_end(usage, latency)`
  callbacks on `Minion`. `chat start --verbose` prints tool calls, token
  usage and per-turn latency to stderr.

**Workspace schema versioning + atomic save**
- Every saved workspace record carries `"schema_version": 1`; a migration hook
  on load treats legacy records as v1, warns on newer/corrupt versions, and is
  the single place future format migrations go. `save_workspaces()` now writes
  atomically via the shared `save_json` helper.

**SQLiteMemory connection lifecycle**
- `SQLiteMemory` supports `with SQLiteMemory(...) as mem:`; the distiller and
  context builder use it instead of manual `try/finally: close()`.

## Why the changes are needed

These are the agent-framework gaps from issue #88 §7: one transient provider
error killed a whole turn, session history grew unbounded into the context,
replies only appeared after the full completion, there was no visibility into
tool calls or token usage, workspace records had no version marker for future
format changes, and DB connections relied on nondeterministic `__del__` cleanup.

## How to validate the changes

`miminions chat start --verbose` — replies stream incrementally, with
`[tool]`/`[turn]` diagnostics on stderr. Existing sessions and `workspaces.json`
files load unchanged.

## Testing

- `pytest tests/unit tests/integration` — 475 passed, no regressions.
- 35 new tests across retry classification/backoff, timeout plumbing, hook
  firing and error isolation, streaming deltas and mid-stream failure,
  history-trim boundary cases, schema-version load/save/migration, and the
  SQLiteMemory context manager.
- `ruff check` clean on all changed files (4 repo lint errors are pre-existing
  in untouched files); mypy introduces no new issues beyond one pre-existing
  false-positive pattern.

### CODE REVIEW

File: tests/unit/test_minion_retry.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+    monkeypatch.setattr("miminions.agent.agent.asyncio.sleep", _record_sleep)
+
+    reply = await minion.run("hello")
+
+    assert reply == "ok"
+    assert delays == [0.5, 1.0]
+
+
+@pytest.mark.parametrize(
```
Comment (shengxio) [inline tests/unit/test_minion_retry.py:108]: This is a really different way of writing the tests. Like it a lot.

File: tests/unit/test_minion_stream.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+async def test_stream_yields_deltas_and_updates_history():
+    minion = create_minion(
+        name="Stream",
+        model=TestModel(custom_output_text="hello world", call_tools=[]),
+    )
+
+    deltas = [delta async for delta in minion.run_stream("hi")]
+
+    assert "".join(deltas) == "hello world"
```
Comment (shengxio) [inline tests/unit/test_minion_stream.py:18]: you may also want to check the deltas are split properly.

File: tests/unit/test_workspace_loading.py [hunk 3]
```diff
@@ -26,3 +31,65 @@ def test_missing_workspaces_file_returns_empty_silently(tmp_path, caplog):
...
+def test_legacy_record_without_schema_version_loads_silently(tmp_path, caplog):
+    """Records saved before versioning existed are treated as v1 — no warning."""
+    legacy = {"ws1": {"id": "ws1", "name": "Legacy", "description": ""}}
+    (tmp_path / "workspaces.json").write_text(json.dumps(legacy))
+    manager = WorkspaceManager(tmp_path)
+
+    with caplog.at_level(logging.WARNING):
+        result = manager.load_workspaces()
+
+    assert result["ws1"].name == "Legacy"
```
Comment (shengxio) [inline tests/unit/test_workspace_loading.py:52]: May also want to assert the legacy workspace will have schema_version == WORKSPACE_SCHEMA_VERSION after the migration.

File: tests/unit/test_workspace_loading.py [hunk 4]
```diff
@@ -26,3 +31,65 @@ def test_missing_workspaces_file_returns_empty_silently(tmp_path, caplog):
...
+
+
+def test_save_stamps_schema_version_and_leaves_no_tmp_file(tmp_path):
+    """save_workspaces writes the version stamp via an atomic replace."""
+    manager = WorkspaceManager(tmp_path)
+    workspace = Workspace(name="Stamped")
+
+    manager.save_workspaces({workspace.id: workspace})
+
+    raw = json.loads((tmp_path / "workspaces.json").read_text())
```
Comment (shengxio) [inline tests/unit/test_workspace_loading.py:63]: Just out of curiousity: any reason why you use read_text + json.loads instead of json.load?

**Review by shengxio — APPROVED** [review]
This PR looks good to me. please resolve the conflict with the changes in this PR. Once done, feel free to merge. Thanks!
